samples/wifi/provisioning/src/main.c

```diff
@@ -9,24 +9,263 @@
+#include <zephyr/bluetooth/gatt.h>
+
+#include <zephyr/net/wifi.h>
+#include <zephyr/net/wifi_mgmt.h>
+#include <net/wifi_credentials.h>
 
 #include "wifi_provisioning.h"
 
+#define ADV_DATA_UPDATE_INTERVAL      5
+
+#define ADV_DATA_VERSION_IDX          (BT_UUID_SIZE_128 + 0)
+#define ADV_DATA_FLAG_IDX             (BT_UUID_SIZE_128 + 1)
+#define ADV_DATA_FLAG_PROV_STATUS_BIT BIT(0)
+#define ADV_DATA_FLAG_CONN_STATUS_BIT BIT(1)
+#define ADV_DATA_RSSI_IDX             (BT_UUID_SIZE_128 + 3)
+
+#define PROV_BT_LE_ADV_PARAM_FAST BT_LE_ADV_PARAM(BT_LE_ADV_OPT_CONNECTABLE, \
+						BT_GAP_ADV_FAST_INT_MIN_2, \
+						BT_GAP_ADV_FAST_INT_MAX_2, NULL)
+
+#define PROV_BT_LE_ADV_PARAM_SLOW BT_LE_ADV_PARAM(BT_LE_ADV_OPT_CONNECTABLE, \
+						BT_GAP_ADV_SLOW_INT_MIN, \
+						BT_GAP_ADV_SLOW_INT_MAX, NULL)
+
+#define ADV_DAEMON_STACK_SIZE 4096
+#define ADV_DAEMON_PRIORITY 5
+
+K_THREAD_STACK_DEFINE(adv_daemon_stack_area, ADV_DAEMON_STACK_SIZE);
+
+static struct k_work_q adv_daemon_work_q;
+
+static uint8_t device_name[] = {'P', 'V', '0', '0', '0', '0', '0', '0'};
+
+static uint8_t prov_svc_data[] = {BT_UUID_PROV_VAL, 0x00, 0x00, 0x00, 0x00};
+
+static const struct bt_data ad[] = {
+	BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	BT_DATA_BYTES(BT_DATA_UUID128_ALL, BT_UUID_PROV_VAL),
+	BT_DATA(BT_DATA_NAME_COMPLETE, device_name, sizeof(device_name)),
+};
+
+static const struct bt_data sd[] = {
+	BT_DATA(BT_DATA_SVC_DATA128, prov_svc_data, sizeof(prov_svc_data)),
+};
+
+static struct k_work_delayable update_adv_data_work;
+static struct k_work_delayable update_adv_param_work;
+
+static void update_wifi_status_in_adv(void)
+{
+	int rc;
+	struct net_if *iface = net_if_get_default();
+	struct wifi_iface_status status = { 0 };
+
+	prov_svc_data[ADV_DATA_VERSION_IDX] = PROV_SVC_VER;
+
+	/* If no config, mark it as unprovisioned. */
+	if (!wifi_prov_state_get()) {
+		prov_svc_data[ADV_DATA_FLAG_IDX] &= ~ADV_DATA_FLAG_PROV_STATUS_BIT;
+	} else {
+		prov_svc_data[ADV_DATA_FLAG_IDX] |= ADV_DATA_FLAG_PROV_STATUS_BIT;
+	}
+
+	rc = net_mgmt(NET_REQUEST_WIFI_IFACE_STATUS, iface, &status,
+				sizeof(struct wifi_iface_status));
+	/* If WiFi is not connected or error occurs, mark it as not connected. */
+	if ((rc != 0) || (status.state < WIFI_STATE_ASSOCIATED)) {
+		prov_svc_data[ADV_DATA_FLAG_IDX] &= ~ADV_DATA_FLAG_CONN_STATUS_BIT;
+		prov_svc_data[ADV_DATA_RSSI_IDX] = INT8_MIN;
+	} else {
+		/* WiFi is connected. */
+		prov_svc_data[ADV_DATA_FLAG_IDX] |= ADV_DATA_FLAG_CONN_STATUS_BIT;
+		/* Currently cannot retrieve RSSI. Use a dummy number. */
+		prov_svc_data[ADV_DATA_RSSI_IDX] = status.rssi;
+	}
+
+	rc = bt_le_adv_update_data(ad, ARRAY_SIZE(ad), sd, ARRAY_SIZE(sd));
+	if (rc != 0) {
+		printk("Cannot update advertisement data, err = %d\n", rc);
+	}
+}
+
+static void connected(struct bt_conn *conn, uint8_t err)
+{
+	char addr[BT_ADDR_LE_STR_LEN];
+
+	if (err) {
+		printk("BT Connection failed (err 0x%02x).\n", err);
```

Comment:
Any specific reason for switching to printk?

Reply:
This part is in user application, so I use printk here